File: tree/binary-search-tree.py
class Node:
    left: 'Node' = None
    right: 'Node' = None
    def __init__(self, value: int) -> None:
        self.value = value

class BinarySeachTree:
    def __init__(self, head: Node) -> None:
        self.head = head

    def __str__(self):
        def printTree(node: Node, level=0) -> list[str]:
            result = []
            if node != None:
                result.extend(printTree(node.right, level + 1))
                result.append(' ' * 4 * level + '-> ' + str(node.value))
                result.extend(printTree(node.left, level + 1))
            return result
        return '\n'.join(printTree(self.head))

    def insert(self, val: int):
        if not self.head:
            self.head = Node(val)
            return
        current = self.head
        while current:
            if current.value < val:
                if not current.right:
                    current.right = Node(val)
                    return
                current = current.right
            elif current.value > val:
                if not current.left:
                    current.left = Node(val)
                    return
                current = current.left
        
    def find(self, value: int) -> Node:
        if not self.head:
            return None
        current = self.head
        while current:
            if value > current.value:
                current = current.right
            elif value < current.value:
                current = current.left
            else:
                return current
        return None

    def min(self, node=None):
        if not node:
            node = self.head
        while node.left != None:
            node = node.left
        return node

    def max(self, node=None):
        if not node:
            node = self.head
        while node.right != None:
            node = node.right
        return node
    
    def delete(self, value):
        if not self.head:
            return
        # Find node
        # Remove
            #   - if node has no child, just remove
            #   - if node has one child, replace it with child
            #   - if node has two children. either:
            #     - replace with the minimum on the right
            #     - replace with the maximum on the left

        def deleteFromNode(value: int, node: Node):
            if value > node.value:
                node.right = deleteFromNode(value, node.right)
                return node
            elif value < node.value:
                node.left = deleteFromNode(value, node.left)
                return node
            else:
                if not node.left and not node.right:
                    return None
                elif node.left != None and not node.right:
                    return node.left
                elif node.right != None and not node.left:
                    return node.right
                else:
                    # Find min on right
                    # Delete
                    # replace
                    minRight = node.right
                    while minRight.left:
                        minRight = minRight.left
                    node.right = deleteFromNode(minRight.value, node.right)
                    node.value = minRight.value
                return node

        deleteFromNode(value, self.head)

    def maxHeight(self) -> int:
        current = self.head
        def height(node: Node) -> int:
            if not node:
                return -1
            maxVal = max(height(node.right), height(node.left))
            return maxVal + 1
        return height(current)

    def minHeight(self) -> int:
        def nodeHeight(node):
            if not node:
                return 0
            left = 1 + nodeHeight(node.left)
            right = 1 + nodeHeight(node.right)
            return min(left, right)
        return nodeHeight(self.head)

    def preOrder(self):
        def traverse(node: Node):
            result: list[int] = []
            if not node:
                return result
            result.append(node.value)
            result.extend(traverse(node.left))
            result.extend(traverse(node.right))
            return result

        return traverse(self.head)

    def inOrder(self):
        def traverse(node: Node):
            result: list[int] = []
            if not node:
                return result
            result.extend(traverse(node.left))
            result.append(node.value)
            result.extend(traverse(node.right))
            return result

        return traverse(self.head)

    def postOrder(self):
        def traverse(node: Node):
            result: list[int] = []
            if not node:
                return result
            result.extend(traverse(node.left))
            result.extend(traverse(node.right))
            result.append(node.value)
            return result

        return traverse(self.head)

    def levelOrder(self):
        queue = []
        result = []
        node = self.head
        while node:
            result.append(node.value)
            if node.left:
                queue.append(node.left)
            if node.right:
                queue.append(node.right)
            node = None
            if len(queue) > 0:
                node = queue.pop(0)

        return result

File: tree/binary-search-tree.go
package trees

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func NewTreeNode(i int) *TreeNode {
	return &TreeNode{Val: i, Left: nil, Right: nil}
}

// ========== Tree Implementation ===============

type BinarySearchTree struct {
	Head *TreeNode
}

func (t *BinarySearchTree) Insert(i int) {
	t.Head = insert(i, t.Head)
}

func (t *BinarySearchTree) Find(i int) *TreeNode {
	return find(i, t.Head)
}

func (t *BinarySearchTree) Min(i int) *TreeNode {
	node := t.Head
	for node.Left != nil {
		node = node.Left
	}
	return node
}

func (t *BinarySearchTree) Max(i int) *TreeNode {
	node := t.Head
	for node.Right != nil {
		node = node.Right
	}
	return node
}

func (t *BinarySearchTree) MaxHeight(i int) int {
	return maxHeight(t.Head)
}

func (t *BinarySearchTree) MinHeight(i int) int {
	return minHeight(t.Head)
}

func (t *BinarySearchTree) Remove(i int) {
	t.Head = deleteNode(i, t.Head)
}

func (t *BinarySearchTree) PreOrder() []int {
	result := make([]int, 0)
	preOrder(t.Head, &result)
	return result
}

func (t *BinarySearchTree) InOrder() []int {
	result := make([]int, 0)
	inOrder(t.Head, &result)
	return result
}

func (t *BinarySearchTree) PostOrder() []int {
	result := make([]int, 0)
	postOrder(t.Head, &result)
	return result
}

func (t *BinarySearchTree) LevelOrder() []int {
	result := make([]int, 0)
	node := t.Head
	stack := []*TreeNode{node}
	for len(stack) > 0 {
		node = stack[0]
		stack = stack[1:]
		result = append(result, node.Val)
		if node.Left != nil {
			stack = append(stack, node.Left)
		}
		if node.Right != nil {
			stack = append(stack, node.Right)
		}
	}
	return result
}

// ========= Utility functions

func insert(i int, node *TreeNode) *TreeNode {
	if node == nil {
		return NewTreeNode(i)
	} else if node.Val > i {
		node.Left = insert(i, node.Left)
		return node
	} else {
		node.Right = insert(i, node.Right)
		return node
	}
}

func find(i int, node *TreeNode) *TreeNode {
	if node == nil {
		return nil
	} else if node.Val == i {
		return node
	} else if node.Val > i {
		return find(i, node.Left)
	} else {
		return find(i, node.Right)
	}
}

func maxHeight(node *TreeNode) int {
	if node == nil {
		return 0
	}
	left := 1 + maxHeight(node.Left)
	right := 1 + maxHeight(node.Right)
	if left > right {
		return left
	}
	return right
}

func minHeight(node *TreeNode) int {
	if node == nil {
		return 0
	}
	left := 1 + maxHeight(node.Left)
	right := 1 + maxHeight(node.Right)
	if left < right {
		return left
	}
	return right
}

func deleteNode(i int, node *TreeNode) *TreeNode {
	if node == nil {
		return nil
	}
	if node.Val > i {
		node.Right = deleteNode(i, node.Right)
		return node
	} else if node.Val < i {
		node.Left = deleteNode(i, node.Left)
		return node
	} else if node.Val == i {
		if node.Left != nil && node.Right != nil {
			// Get min at right
			minNode := node.Right
			for minNode.Left != nil {
				minNode = minNode.Left
			}
			// delete it
			node.Right = deleteNode(minNode.Val, node.Right)
			// replace current node with min node
			node.Val = minNode.Val
			return node
		} else if node.Right != nil {
			return node.Right
		} else if node.Left != nil {
			return node.Left
		}
	}
	return nil
}

func preOrder(node *TreeNode, result *[]int) {
	if node == nil {
		return
	}
	*result = append(*result, node.Val)
	preOrder(node.Left, result)
	preOrder(node.Right, result)
}

func inOrder(node *TreeNode, result *[]int) {
	if node == nil {
		return
	}
	inOrder(node.Left, result)
	*result = append(*result, node.Val)
	inOrder(node.Right, result)
}

func postOrder(node *TreeNode, result *[]int) {
	if node == nil {
		return
	}
	postOrder(node.Left, result)
	postOrder(node.Right, result)
	*result = append(*result, node.Val)
}
